docs/proposals/push-metrics.md

```diff
@@ -0,0 +1,69 @@
+# Heapster Push Metrics
+
+## Overview and Motivation
+
+Currently, Heapster supports pulling metrics from kubelet, and defines an
+interface for pulling from other sources.  However, in certain cases, it is
+more useful to be able to have services push metrics into Heapster, instead
+of having Heapster pull metrics.
+
+For instance, supporting push metrics makes it easy for cluster admins to add
+custom metrics into Heapster with relatively minimal effort: they can simply
+write a program or script which collects and processes the information, and
+then add a recurring Job or cron job which pushes the metrics (instead of
```

Comment:
You may wish to take a look at how the Prometheus Node Exporter Textfile Collector works as an alternate implementation option. It's how we handle machine level metrics that aren't produced at every scrape, which seems to be the scenario here and has worked quite well for us.
Basically it's a directory you can dump prometheus text format files into, and they'll be included in `/metrics` output.

In this case the kubelet could pull in files on scrape, and include them with the rest of it's metrics. Depending on directory layout, file permissions and label handling this could remove the need to do explicit auth, users having to figure out which container/pod label to apply and opens options for non-kubernetes systems on the machine to produce metrics.

For things you need to set only once this would avoid the need to setup a cronjob or depend on the scheduler to run one (and be in a sufficiently working state to be able to schedule). It also avoids the need for rate limiting, as it's only a write to local disk rather than hitting heapster directly.